[Sema] Warn on empty catch blocks in do-catch statements

AnirudhMathur12 opened the pull request:
This adds a diagnostic to warn when a catch block is empty.

This pull request introduces a new semantic diagnostic to warn when a catch block in a do-catch statement is empty. Swallowing errors silently without explicit acknowledgement can lead to bugs that are difficult to trace.

While Swift allows empty catch blocks, making the intent explicit improves code safety. This implementation suggests a Fix-it to insert _ = error, which is the idiomatic way to acknowledge an ignored error in Swift.

#### Key Changes:
- Added empty_catch_block warning to DiagnosticsSema.def.
- Implemented checking logic in MiscDiagnostics.cpp within the performStmtDiagnostics pass.
- Added logic to skip the warning if the user provides an explicit acknowledgement (i.e., _ = error).
- Added comprehensive regression tests in test/Sema/diag_empty_catch.swift.

Resolves #43590

AnirudhMathur12 committed:
[Sema] Warn on empty catch blocks in do-catch statements

This adds a diagnostic to warn when a catch block is empty and provides a fix-it to explicitly ignore the error with '_ = error'.

MaxDesiatov commented on test/Sema/diag_empty_catch.swift:
```diff
@@ -0,0 +1,25 @@
+  } catch {
+    print("Logged")
+  }
+}
```
A trailing newline is missing here, as visible in the GitHub diff viewer.

MaxDesiatov commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -59,8 +59,8 @@ using namespace constraints;
-    if (auto IIOE = dyn_cast<InjectIntoOptionalExpr>(
-                                               E->getSemanticsProvidingExpr()))
+    if (auto IIOE =
+            dyn_cast<InjectIntoOptionalExpr>(E->getSemanticsProvidingExpr()))
```
Formatting changes unrelated to the fix such as this one should be omitted or extracted into a separate PR.

jamieQ commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -59,8 +59,8 @@ using namespace constraints;
 /// Return true if this expression is an implicit promotion from T to T?.
```
FYI @AnirudhMathur12, regarding the unintended reformatting here – there is a [section in the FAQ](https://github.com/swiftlang/swift/blob/652402e9e83d2e970de5a12c34e419f95d1cb7f3/docs/HowToGuides/FAQ.md#how-do-i-format-my-changes) regarding how to use `git-clang-format` to only reformat the altered portion of files in a git changeset. that tool works well for the c++ code in this codebase.

AnirudhMathur12 commented:
Hi, @xwu 
I'm away from my machine on vacation and will get back to it in a couple of days. Thank you so much for your advice and your patience!

AnirudhMathur12 commented:
Hi @xwu and @MaxDesiatov,
I have made the changes you all requested.
- Added the extra test case which was originally mentioned in the issue.
- Reverted all the clang-format changes
- Added trailing newline at the end of the test file.
Let me know if any adjustments are needed. Thanks in advance!

jamieQ commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -9060,5 +9060,7 @@ ERROR(unsafe_self_dependent_result_attr_on_invalid_decl,none,
 
 REMARK(macro_expansion_line, none, "macro content: |%0|", (StringRef))
 
+WARNING(empty_catch_block, none, "catch block is empty, errors will silently be ignored", ())
```
nit: a comma feels slightly wrong here... maybe a semicolon would be more fitting? @xwu any thoughts on this?

AnirudhMathur12 replied:
I have changed it to "empty catch block silently ignores errors; consider using 'try?' or 'catch _' to explicitly ignore".

xwu replied:
Nits: (1) ideally, I think, you’d add this line to the file alongside the other diagnostics about do…catch and not just at the end of the file; (2) re the wording, I don’t think the block is doing the ignoring; perhaps it is right to say that it silences all errors, but that is also at best just restating what the code does in English, so I’m also not sure that it’s much helpful—writing is tough!; (3) don’t we have a policy now that all new diagnostics should go into a named diagnostic group?

AnirudhMathur12 replied:
Closing the review as the topics were brought up later on and discussed there.

AnirudhMathur12 commented on test/Sema/diag_empty_catch.swift:
```diff
@@ -0,0 +1,32 @@
+
+func canThrow() throws {}
+
+func test() {
```
noted.

AnirudhMathur12 replied:
Updated.

AnirudhMathur12 commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1418,7 +1418,7 @@ ERROR(if_expr_must_be_syntactically_exhaustive,none,
       ())
 ERROR(do_catch_expr_must_be_syntactically_exhaustive,none,
-      "'do catch' must have an unconditional 'catch' to be used as expression",
+      "'do catch' must have an unconditional '  catch' to be used as expression",
```
noted.

AnirudhMathur12 replied:
Fixed.

AnirudhMathur12 committed:
[Sema] Enforce strict diagnostics for empty catch blocks

This patch implements stricter diagnostics for empty `do-catch` blocks to prevent accidental error swallowing.

harlanhaskins commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -9060,5 +9060,8 @@ ERROR(unsafe_self_dependent_result_attr_on_invalid_decl,none,
 
 REMARK(macro_expansion_line, none, "macro content: |%0|", (StringRef))
 
+WARNING(empty_catch_block, none, "empty catch block silently ignores errors; consider using 'try?' or 'catch _' to explicitly ignore", ())
```
tiny nitpick: I think I'd rather we not encourage developers to reach for `try?` when they've already gone through the hassle of wrapping in a do/catch, personally. Other folks in the community may feel differently.

harlanhaskins commented:
Well, there's nothing special about specifically `_ = error` as a syntactic construct, any discard-assignment would technically be a valid way to silence the warning, or a call to an empty function, or a while false, or any number of possible statements.

AnirudhMathur12 commented:
> I do agree with the change to silence this via `catch _`, as it seems more obvious, but also I don't think we now need a diagnostic to recognize and point out a syntactic construct that originated as the previous silencing mechanism for this PR. Just recognizing the empty block and suggesting the `_` is fine.

I added the redundancy check because writing `_ = error` is correct syntax, but it is not the "preferred" one. Another problem I had was that there would be two correct ways to address the same problem in code, which is not a big problem but would be inappropriate from an idiomatic perspective.

harlanhaskins commented:
Yes, but `_ = error` is only relevant to this PR because it was previously proposed in this PR as the preferred suppression method. There's nothing special about `_ = error` as opposed to `if true {}` or any other statement. There are infinite ways to silence this warning.

xwu commented on test/Sema/diag_empty_catch.swift:
```diff
@@ -0,0 +1,52 @@
+    try something()
+  } catch MyError.bad {
+    print("bad")
+  } catch _ {
```
```suggestion
  } catch {
```

No, the test case I mentioned from the original ticket that should pass must allow bare ‘catch’ here.

AnirudhMathur12 replied:
I understand the reference to the original ticket, but I believe allowing a bare `catch { }` here would defeat the primary purpose of this PR.

Technically, a bare catch implicitly captures the error (`let error`), so an empty body effectively swallows that captured error silently. The goal of this change is to require the user to be intentional about ignoring errors by using `catch _`.

I am happy to add the test case you requested, but under these new strict rules, it will trigger a warning. I will add it as a test case that expects the empty_catch_block warning.

Does that sound correct, or is there a specific reason we should exempt bare catches in this context?

xwu replied:
Sorry, no--if the intent is to forbid bare `catch { }` altogether, I think that would be tantamount to a language-level change. That's explicitly not what the original bug report is about, as the commenters there clarified.

AnirudhMathur12 replied:
I see. I will address that.

AnirudhMathur12 committed:
[Sema] Stop warning on explicit error silencing via assignment

The compiler previously emitted a "redundant error capture" warning for
`catch { _ = error }`, suggesting `catch _` instead. This change removes
that restriction, treating explicit assignment to `_` as a valid way to
handle the error.

Also refactors the `empty_catch_block` diagnostic definition.

AnirudhMathur12 committed:
[Diagnostics] Warn on empty catch blocks only when unchained

Introduces a warning for empty `catch` blocks that silently swallow errors,
but restricts it to `do` statements with a single catch clause.

- Warns on `do { ... } catch { }` (blindly swallowing errors).
- Allows `do { ... } catch MyErr { ... } catch { }` (ignoring remainder).
- Allows explicit ignore via `catch _`.

AnirudhMathur12 commented:
**Update:** Addressed feedback from @xwu regarding chained catch blocks.

I updated the logic to check `doCatch->getCatches().size()`. The compiler now **only** warns if the `do` statement has a single catch block that is empty. 

If an empty catch is part of a chain (e.g., `catch MyErr { ... } catch { }`), it is now considered valid and does not emit a warning.

xwu commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1420,6 +1420,11 @@ ERROR(if_expr_must_be_syntactically_exhaustive,none,
       "'do catch' must have an unconditional 'catch' to be used as expression",
       ())
+
+WARNING(empty_catch_block,NoUsage,
```
We should probably workshop the name of this warning group, unless it's a pre-existing one. Otherwise, "NoUsage" lacks context and is probably too general. (Naming is hard, though, so I'm struggling to propose a better name.)

xwu replied:
At some point, it was said that every new warning added should have a diagnostic group--I'm not sure if that's been kept up with but we can be good citizens and adopt a good name for the group here.

To riff on your suggestion, "EmptyUnconditionalCatch" or something like that?

DougGregor replied:
New warnings really do need to go into a diagnostic group. What kinds of problems are you hitting?

AnirudhMathur12 replied:
I am unable to create a new warning group. I added `GROUP(EmptyUnconditionalCatch, none, "empty-unconditional-catch")` to `DiagnosticGroups.def`. But after recompiling it would still tell me that the group doesn't exist. I might be doing something wrong. I am not familiar with making groups.

xwu reviewed (commented):
Minor nits from my perspective, but this is looking good!  
Thanks for sticking to it.

AnirudhMathur12 committed:
[Sema] Refine empty catch warning text and remove Fix-It

Addresses code review feedback suggested by @xwu:
- Updates the diagnostic message to be more explicit that "all errors" are silenced.
- Clarifies the suggestion to use 'try?' or explicit 'catch _'.
- Removes the automatic Fix-It that inserted '_', leaving the resolution up to the user.

jamieQ commented on test/Sema/diag_empty_catch.swift:
```diff
@@ -0,0 +1,53 @@
+  // 1. Standard Empty Catch -> WARN
+  do {
+    try something()
+  } catch { // expected-warning {{empty catch block silences errors; use 'try?' or 'catch _' to explicitly ignore}}
```
nit: i assume these haven't been updated yet due to iterating on the diagnostic wording

AnirudhMathur12 replied:
Thanks for catching that mistake! I have addressed it and will be pushing soon.

AnirudhMathur12 commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -6521,6 +6521,29 @@ void swift::performStmtDiagnostics(const Stmt *S, DeclContext *DC) {
+      auto *pattern = catchStmt->getCaseLabelItems().front().getPattern();
+      if (isa<AnyPattern>(pattern->getSemanticsProvidingPattern())) continue;
+
+      Identifier boundErrorName;
```
Addressed.

AnirudhMathur12 commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -6521,6 +6521,29 @@ void swift::performStmtDiagnostics(const Stmt *S, DeclContext *DC) {
   if (!ctx.LangOpts.DisableAvailabilityChecking)
     diagnoseStmtAvailability(S, const_cast<DeclContext*>(DC));
+
+  if (auto *doCatchStmt = dyn_cast<DoCatchStmt>(S)) {
```
I separated it out into another function, but I wasn't sure about the function's positioning so I put it just before `performStmtDiagnostics`. Please let me know if there is a better place to put the code.

jamieQ replied:
personally, i'd probably move it next to `diagnoseCaseStmtAmbiguousWhereClause`. currently it's under the area commented as 'high-level entry points', which it is not. also maybe rename to `diagnoseEmptyCatchBlock` to better match existing conventions.

AnirudhMathur12 commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -6521,6 +6521,29 @@ void swift::performStmtDiagnostics(const Stmt *S, DeclContext *DC) {
+      if (!body) continue;
+
+      auto *pattern = catchStmt->getCaseLabelItems().front().getPattern();
+      if (isa<AnyPattern>(pattern->getSemanticsProvidingPattern())) continue;
```
Addressed.

AnirudhMathur12 committed:
[Sema] Refactor empty catch diagnostic logic

Extracts the empty catch block diagnostic into a dedicated helper function, `checkEmptyCatchBlock`, for better encapsulation and readability.

Improvements include:
- Optimizes the check to exit early if the `do-catch` statement does not have exactly one catch clause.
- Uses `hasCaseBodyVariables()` to reliably distinguish between catch blocks that implicitly or explicitly bind variables (which should warn if empty) and explicit ignores like `catch _` (which should not warn).
- Updates test expectations to match the refined diagnostic message.

AnirudhMathur12 commented:
I’ve updated the PR to address the latest feedback by @jamieQ. The logic is now encapsulated in a `checkEmptyCatchBlock` helper function, which uses `hasCaseBodyVariables()` for a more robust check and includes the early exit optimization. Thanks for the guidance!

jamieQ commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -6506,6 +6506,31 @@ void swift::performSyntacticExprDiagnostics(const Expr *E,
+    return;
+
+  auto *body = dyn_cast_or_null<BraceStmt>(catchStmt->getBody());
+  if (body && body->empty()) {
```
nit: `getBody()` returns a `BraceStmt*`, so the cast here seems superfluous (though please correct me if i'm missing something):

```suggestion
  auto *body = catchStmt->getBody();
  if (body && body->empty()) {
```

or sometimes i see code where the variable is bound within the `if` to scope it to the condition like:

```cxx
if (auto *body = catchStmt->getBody();
    body && body->empty()) {
```

but i think either are probably fine here since the method body is pretty short.

AnirudhMathur12 committed:
[Sema] Simplify empty catch check logic

Removes a redundant dynamic cast since `CatchStmt::getBody()` already returns a `BraceStmt`. Also adopts the scoped-if syntax for the body variable to tighten the scope, addressing code review feedback.

xwu commented:
@swift-ci smoke test

xwu commented:
Hmm, it's unexpected that there are dozens of failures from this one diagnostic change, no? This raises questions for investigation:

Does the test suite resemble real-world code? If so, are we worried about introducing dozens of new warnings to people's working code bases for one nice-to-have warning?

Do the test failures show that warnings are appearing only in the narrowly tailored cases we want to warn about, or are the warnings flagging usages they shouldn't be flagging?

In general, one shouldn't change unrelated test cases to pass, but neither should one just mechanically add warning expectations. In each case it's a case-by-case decision as to what _should_ happen.

AnirudhMathur12 committed:
[Sema] Refine empty catch warning to avoid false positives and update test cases

This patch tightens the logic for the empty catch diagnostic to prevent it from triggering on valid code patterns or test scaffolding and replaces catch {} with catch _ {} in some cases.

The warning is now suppressed if:
- The catch block has a guard (`where` clause), as it is conditional and does not silence all errors.
- The catch pattern is explicit (e.g., `catch MyError`), focusing the warning strictly on implicit `catch {}`.
- The catch pattern itself contains compiler errors, to avoid adding noise to code that is already broken.

Additionally, updated a few test case files, where catch {} was replaced
catch _ {}

AnirudhMathur12 committed:
No warning emitted when the catch is unreachable

AnirudhMathur12 commented on test/decl/func/debugger_function.swift:
```diff
@@ -50,15 +50,15 @@ func test2() -> Int {
-    catch {} // expected-warning {{'catch' block is unreachable because no errors are thrown in 'do' block}}
-  } catch {
+    catch _ {} // expected-warning {{'catch' block is unreachable because no errors are thrown in 'do' block}}
+  } catch _ {
```
Addressed

xwu replied:
Since you've addressed this logic, these changes should be reverted now.

AnirudhMathur12 commented on test/SILOptimizer/consume_operator_kills_copyable_loadable_vars.swift:
```diff
@@ -216,7 +216,7 @@ struct S {
             let maybeNewB = exchangeUse(b)
             self = .init(buffer: maybeNewB)
-        } catch {
+            } catch {
```
Reverted.

AnirudhMathur12 committed:
[Sema] Improve empty catch diagnostics with deep throw detection

Previously, the "empty catch block silences all errors" warning was "lazy": it only checked for throwing expressions at the top level of the do-block. This caused false negatives where `try` expressions or explicit `throw` statements were nested inside control flow (e.g., loops, if/else, guard) or nested `do-catch` blocks.

This patch:
- Implements a recursive AST search in `checkEmptyCatchBlock` to correctly detect throwing statements nested arbitrarily deep within the `do` body.
- Adds specific handling for `ThrowStmt` to ensure explicit `throw` statements trigger the warning (previously ignored).
- Registers the new `EmptyUnconditionalCatch` warning group in `DiagnosticGroups.def`.
- Updates `diag_empty_catch.swift` with comprehensive test cases covering recursion, control flow, explicit throws, and nested re-throws.
- Fixed indentation in consume_operator_kills_copyable_loadable_vars.swift

AnirudhMathur12 commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -6552,17 +6552,80 @@ static void checkEmptyCatchBlock(const DoCatchStmt *doCatchStmt, ASTContext &ctx
   }
 
-  // Use scoped-if to keep 'body' local to the check.
+  if (!bodyThrows)
```
I checked `lib/Sema/TypeCheckEffects.cpp`. The logic that detects unreachable catch blocks corresponds to the `diag::no_throw_in_do_with_catch` diagnostic.

However, that logic is tightly coupled to the `CheckEffectsCoverage` class (defined in an anonymous namespace), which tracks state like `ContextFlags::HasAnyThrowSite` during a full walk of the function body.

Since `CheckEffectsCoverage` is file-private exposing a lightweight public API just to check 'does this statement throw' would require a significant refactor out of the scope of this PR. I believe the recursive helper in `MiscDiagnostics.cpp` is adequate for this PR.

I will separate out the logic into a separate function.

AnirudhMathur12 committed:
Refactored the no throwing do detection into its own function

AnirudhMathur12 committed:
[Sema] Diagnose empty catch blocks using effects analysis

Previously, detecting whether an empty catch block silenced an error was not supported. Initial attempts to implement this via a custom AST walker were prone to missing complex control flow, explicit throw statements, or inactive code blocks (e.g., #if).

This patch implements the `EmptyUnconditionalCatch` warning by integrating directly into the `CheckEffectsCoverage` pass in `TypeCheckEffects.cpp`.

Changes:
- Adds `diagnoseEmptyCatch` to `CheckEffectsCoverage`.
- Hooks into `checkExhaustiveDoBody` and `checkNonExhaustiveDoBody` to reuse the existing `ContextFlags::HasAnyThrowSite` analysis.
- This ensures the warning only triggers if the `do` block actually throws, automatically handling edge cases like explicit throws, nested re-throws, and conditional compilation (#if).
- Adds comprehensive regression tests in `diag_empty_catch.swift`, covering recursion, control flow, and conditional compilation.
- Reverts any changes made to MiscDiagnostics.cpp

AnirudhMathur12 committed:
Minor fixes and test case improvements

- Removed unnecessary placement of _ in test/Parse/errors.swift.
- Moved empty catch check higher for early exit
- Warns on explicit bindings what are not exclusively `_`

AnirudhMathur12 committed:
Updated test cases to expect the empty catch warning

AnirudhMathur12 committed:
[Sema] Fix 'empty catch' warning firing on specific patterns and 'is Error'

Previously, `diagnoseEmptyCatch` would warn on any empty catch block that wasn't explicitly `catch _`. This caused two issues:
1. It triggered false positives for specific error matches (e.g., `catch .failed`, `catch _ as NSError`), which are valid ways to swallow specific errors.
2. It created redundant warnings for `catch is Error`, which now has a dedicated diagnostic ("'is' test is always true").

This change refines the logic to inspect the underlying semantic pattern. The warning is now only emitted for confirmed catch-all patterns (e.g., `catch let error` or `catch let e as Error`), while correctly ignoring specific patterns and `catch is Error`.

AnirudhMathur12 committed:
Reverted unintended whitespace changes and removed unnecessary modification of a test case
